apps/bfd-pipeline/bfd-pipeline-ccw-rif/src/test/java/gov/cms/bfd/pipeline/ccw/rif/extract/RifFileParserTest.java

```diff
@@ -0,0 +1,209 @@
+   */
+  private final RifFile rifFile = new LocalRifFile(tempFile, RifFileType.BENEFICIARY);
+
+  /** We need this when we create a {@link RifRecordEvent} but we never call any of its methods. */
+  @Mock private RifFileEvent rifFileEventMock;
+
+  /**
+   * Creates the temp file before any tests have run.
+   *
+   * @throws IOException pass through
+   */
+  @BeforeAll
+  static void beforeAll() throws IOException {
+    tempFile = Files.createTempFile(RifFileParser.class.getName(), ".csv");
+  }
+
+  /**
+   * Deletes the temp file after all tests have run.
+   *
+   * @throws IOException pass through
+   */
+  @AfterAll
+  static void afterAll() throws IOException {
+    if (tempFile != null) {
+      Files.deleteIfExists(tempFile);
+    }
+  }
+
+  /**
+   * Verifies edge conditions for {@link RifFileParser.Simple}.
+   *
+   * @throws IOException pass through from writing string to temp file
+   */
+  @Test
+  void simpleAlwaysUsesOneRecordPerEvent() throws IOException {
+    final var parser = new RifFileParser.Simple(this::parseSingle);
+    // empty file should produce no records
+    assertEquals(List.of(), parseString("", parser));
+
+    // single record in file should work
+    assertEquals(List.of("1-a"), parseString("id|data\n1|a\n", parser));
+
+    // several single records should work
+    assertEquals(List.of("1-a", "2-a", "3-a"), parseString("id|data\n1|a\n2|a\n3|a\n", parser));
+
+    // several with mixed number of records should work
+    assertEquals(
+        List.of("1-a", "2-a", "2-b", "3-a", "3-b", "3-c"),
+        parseString("id|data\n1|a\n2|a\n2|b\n3|a\n3|b\n3|c\n", parser));
+  }
+
+  /**
+   * Verifies edge conditions for {@link RifFileParser.Grouping}.
+   *
+   * @throws IOException pass through from writing string to temp file
+   */
+  @Test
+  void groupingHonorsIdColumn() throws IOException {
+    final var parser = new RifFileParser.Grouping("id", this::parseGroup);
+    // empty file should produce no records
+    assertEquals(List.of(), parseString("", parser));
+
+    // single record in file should work
+    assertEquals(List.of("1-a"), parseString("id|data\n1|a\n", parser));
+
+    // several single records should work
+    assertEquals(List.of("1-a", "2-a", "3-a"), parseString("id|data\n1|a\n2|a\n3|a\n", parser));
+
+    // several with mixed number of records should work
+    assertEquals(
+        List.of("1-a", "2-a;2-b", "3-a;3-b;3-c"),
+        parseString("id|data\n1|a\n2|a\n2|b\n3|a\n3|b\n3|c\n", parser));
+
+    // missing grouping column should throw an exception
+    assertThrows(RuntimeException.class, () -> parseString("noid|data\n1|a\n2|a\n3|a\n", parser));
+  }
+
+  /**
+   * Similar to {@link #groupingHonorsIdColumn} but uses randomly generated rif data to test wider
+   * variety of scenarios {@link RifFileParser.Grouping}.
+   *
+   * @throws IOException pass through from writing string to temp file
+   */
+  @Test
+  void groupingHonorsIdColumnRandomScenarios() throws IOException {
```

Comment:
This is completely predictable.  The seed is fixed so the same sequence of test cases will be repeated every time this test runs.  If it did fail it because of a java.util.Random change (following a JDK version upgrade) then we could respond to it.  That's unlikely to happen more than once a year at most.